pennylane/labs/resource_estimation/resource_tracking.py

```diff
@@ -0,0 +1,316 @@
+        wires: 3
+        gates: 6
+        gate_types:
+        {'Hadamard': 2, 'CNOT': 1, 'RX': 1, 'RY': 1, 'QFT': 1}
+
+        We can also obtain resources for individual operations and quantum tapes in a similar manner:
+
+        >>> op = re.ResourceRX(1.23, 0)
+        >>> print(re.get_resources(op))
+        wires: 1
+        gates: 17
+        gate_types:
+        {'T': 17}
+
+        Finally, we can modify the config values listed in the global :code:`labs.resource_estimation.resource_config`
+        dictionary to have finegrain control of how the resources are computed.
+
+        >>> re.resource_config
+        {'error_rx': 0.01, 'error_ry': 0.01, 'error_rz': 0.01}
+        >>>
+        >>> my_config = copy.copy(re.resource_config)
+        >>> my_config["error_rx"] = 0.001
+        >>>
+        >>> print(re.get_resources(op, config=my_config))
+        wires: 1
+        gates: 21
+        gate_types:
+        {'T': 21}
+
+    """
+
+    raise TypeError(
+        f"Could not obtain resources for obj of type {type(obj)}. obj must be one of Operation, Callable or QuantumScript"
+    )
+
+
+@get_resources.register
+def resources_from_operation(
+    obj: Operation, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Resources:
+    """Get resources from an operation"""
+
+    if isinstance(obj, ResourceOperator):
+        cp_rep = obj.resource_rep_from_op()
+
+        gate_counts_dict = defaultdict(int)
+        _counts_from_compressed_res_op(cp_rep, gate_counts_dict, gate_set=gate_set, config=config)
+        gate_types = _clean_gate_counts(gate_counts_dict)
+
+        n_gates = sum(gate_types.values())
+        return Resources(num_wires=len(obj.wires), num_gates=n_gates, gate_types=gate_types)
+
+    res = Resources()  # TODO: Add implementation here!
+    return res
+
+
+@get_resources.register
+def resources_from_qfunc(
+    obj: Callable, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Callable:
+    """Get resources from a quantum function which queues operations!"""
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        operations = tuple(op for op in q.queue if not isinstance(op, MeasurementProcess))
+        compressed_res_ops_lst = _operations_to_compressed_reps(operations)
+
+        initial_gate_set = set.union(gate_set, _StandardGateSet)
+
+        gate_counts_dict = defaultdict(int)
+        for cmp_rep_op in compressed_res_ops_lst:
+            _counts_from_compressed_res_op(
+                cmp_rep_op, gate_counts_dict, gate_set=initial_gate_set, config=config
+            )
+
+        # Validation:
+        condensed_gate_counts = defaultdict(int)
+        for sub_cmp_rep, counts in gate_counts_dict.items():
+            _counts_from_compressed_res_op(
+                sub_cmp_rep, condensed_gate_counts, scalar=counts, gate_set=gate_set, config=config
+            )
```

Comment:
In the first loop we get the counts for every gate in the union of `gate_set` and `_StandardGateSet`, then in the second we decompose the gates from `_StandardGateSet` into gates from `gate_set`, is this correct? Why can't we do this with one loop and just directly decompose into gates from `gate_set`?

Reply:
This is an optimization based on the following observation: 

- Most algorithms will be decomposed into ~ O(100,000) or more gates from the set {X, Y, Z, SWAP, RX, RY, RZ, PhaseShift}. <-- Also probably Controlled versions of these gates but I exclude them for now. 
- The conversion from that layer to the final gateset (which is usually) {Hadamard, T, S, CNOT, Toffoli}, is a static function call 

So the idea is instead of calling the static functions O(100,000) times, we instead just count up the gates at that intermediate stage already and then only do a single function call and scale accordingly.